src/rc_string.c

```diff
@@ -7,4 +7,168 @@
 #include "azure_c_pal/gballoc.h"
 #include "azure_c_util/rc_string.h"
 #include "azure_c_logging/xlogging.h"
+#include "azure_c_util/thandle.h"
 
+#define THANDLE_MALLOC_FUNCTION malloc
+#define THANDLE_FREE_FUNCTION free
+THANDLE_TYPE_DEFINE(RC_STRING);
+
+#define STRING_STORAGE_TYPE_VALUES \
+    STRING_STORAGE_TYPE_COPIED, \
+    STRING_STORAGE_TYPE_MOVED, \
+    STRING_STORAGE_TYPE_WITH_CUSTOM_FREE
+
+MU_DEFINE_ENUM(STRING_STORAGE_TYPE, STRING_STORAGE_TYPE_VALUES)
+
+// In order not to expose the storage type to the user (as it is something internal to the implementation),
+// we're going to have another structure that wraps the RC_STRING and the storage type
+// It is imperative that the first member is the RC_STRING structure though, so it matches the strcuture that
+// we return to the user
+typedef struct RC_STRING_INTERNAL_TAG
+{
+    RC_STRING rc_string;
+    STRING_STORAGE_TYPE storage_type;
+    RC_STRING_FREE_FUNC free_func;
+    void* free_func_context;
+} RC_STRING_INTERNAL;
+
+#define RC_STRING_INTERNAL_FROM_RC_STRING(rc_string_content) \
+    (RC_STRING_INTERNAL*)(void*)(rc_string_content);
```

Comment:
>rc_string_content [](start = 33, length = 17)

I like the `offsetof` code here (or I guess CONTAINING_RECORD) instead because then there isn't some implicit assumption that the `RC_STRING` is the first member. It's not obvious that you can't change its position in `RC_STRING_INTERNAL`.